GPS WPF tutorial (#457)
ros-planning/navigation.ros.org: tutorials/docs/navigation2_with_gps.rst

> .. _navigation2-with-gps:

Reviewer: This tutorial wasn't added to the index.html so it doesn't appear on the website :-)

> Navigating with a GPS Based Localization System

Reviewer: How about just "Navigating Using GPS Localization"

Make sure to adjust the `****` to be the same length under

GPS devices calculate their position using the `WGS84 standard <https://en.wikipedia.org/wiki/World_Geodetic_System>`_, which defines a cartesian system the earth's origin on its center of mass, the `z` axis pointing north and the `x` axis pointing to the first meridian as the image below shows.

.. image:: images/Gps_Navigation/WGS_84_reference_frame.svg
    :width: 562px
    :align: center
    :alt: WGS84 reference frame

However, this reference system is impractical for describing the motion of objects in or close to the earth's surface: Imagine your robot is located on a soccer field and you want it to move from one end to the other, your navigation task would look something like:

  "go from X=4789.413km, Y=177.511km z=4194.292km to X=4789.475km, Y=177.553km z=4194.22km"
  
It would make much more sense to create a local reference system where you could tell your robot "go 100 meters forward", right?

> To cope with this, geodesy allows for the definition of  `Local tangent planes <https://en.wikipedia.org/wiki/Local_tangent_plane_coordinates>`_. According to `REP 103 <https://www.ros.org/reps/rep-0103.html>`_ the coordinate systems of these planes should follow the ENU convention, which means their +x axis should face East, +y axis should face North and the +z axis should point Up, as the image below shows. In the context of this tutorial the GPS coordinates of the origin of the local cartesian plane will be called **datum** as it is in `robot_localization <http://docs.ros.org/en/noetic/api/robot_localization/html/index.html>`_

Reviewer: ```suggestion
To cope with this, geodesy allows for the definition of  `Local tangent planes <https://en.wikipedia.org/wiki/Local_tangent_plane_coordinates>`_. According to `REP 103 <https://www.ros.org/reps/rep-0103.html>`_ the coordinate systems of these planes should follow the ENU convention, which means their +x axis should face East, +y axis should face North and the +z axis should point Up, as the image below shows. In the context of this tutorial the GPS coordinates of the origin of the local cartesian plane will be called **datum** as it is in `robot_localization <http://docs.ros.org/en/noetic/api/robot_localization/html/index.html>`_.
```

    :width: 520px
    :align: center
    :alt: Local tangent plane

In this tutorial we assume the robot's GPS produces a really accurate and smooth estimation of the robot's position, however in the real world for standalone GPSs that's often not the case: you should expect accuracies of 1-2 meters under excellent conditions and up to 10 meters, and frequent jumps in the position as the GPS sensor picks up less or more satellites.

Several positioning augmentation technologies exists to reduce the error of GPS measurements, one of the most common ones is called `RTK <https://en.wikipedia.org/wiki/Real-time_kinematic_positioning>`_ (Real Time Kinematic Positioning), which can bring the accuracy of receivers down to 1cm. If Accuracy matters in your application this technology is highly recommended; though this requires the deployment of a second fixed GPS called Base, most of the US and Europe are already covered with public free to use Bases that you can connect to. You can read more about RTK and how to get started on `this tutorial <https://learn.sparkfun.com/tutorials/setting-up-a-rover-base-rtk-system>`_.

> Additionally, to fully describe a robot's localization using the ENU convention we need to know its heading as well, however GPS sensors do not provide orientation measurements, only position measurements. Though there are other alternatives for measuring orientation, in this tutorial we assume the robot is equipped with an IMU capable of providing absolute heading measurements, meaning it will output zero yaw when facing east and +90 degrees when facing north. 

Reviewer: Is that a realistic expectations of IMUs? Don't they typically start at 0 when powered on / initialized?  That's not really what is meant by "absolute" when you buy IMUs (it just means it has a way to fuse all the IMU sensors to get an absolute orientation, but its not necessarily aligned to anything but the initial pose) 

I know you talk about that a bit in the later paragraphs about how that can't be relied on, but I don't think this is actually a true statement. I think this should be adjusted.

Despite the above assumption, in the real world commercial grade IMU's mounted in actual robots will often not produce accurate measurements of absolute heading because: 

1. They are hard to calibrate: outdoors robots are often big and heavy: imagine doing an eight figure in the air with an autonomous tractor.

2. They rely on magnetometers to measure earth's magnetic field but robots are often a huge source of electromagnetic noise: Electric motors are full of permanent magnets and can draw several amps, producing significant disturbances to the sensor.

Through the development of the tutorial we will see how to leverage robot_localization's Kalman Filters to mitigate this problem.

Tutorial Steps
==============

0- Setup Gazebo World
---------------------

> To navigate using GPS we first need to create an outdoors Gazebo world with a robot having a GPS sensor. For this tutorial we will be using the `sonoma raceway <https://docs.px4.io/main/en/sim_gazebo_classic/gazebo_worlds.html#sonoma-raceway>`_ because its aligned with the real location. A sample world has been setup `here <https://github.com/ros-planning/navigation2_tutorials/tree/master/nav2_gps_waypoint_follower_demo/worlds/sonoma_raceway.world>`_ using gazebo's spherical coordinates plugin, which creates a local tangent plane centered in the set geographic origin and provides latitude, longitude and altitude coordinates for each point in the world.

Reviewer: ```suggestion
To navigate using GPS we first need to create an outdoors Gazebo world with a robot having a GPS sensor to setup for navigation. For this tutorial we will be using the `sonoma raceway <https://docs.px4.io/main/en/sim_gazebo_classic/gazebo_worlds.html#sonoma-raceway>`_ because its aligned with the real location. A sample world has been setup `here <https://github.com/ros-planning/navigation2_tutorials/tree/master/nav2_gps_waypoint_follower_demo/worlds/sonoma_raceway.world>`_ using gazebo's spherical coordinates plugin, which creates a local tangent plane centered in the set geographic origin and provides latitude, longitude and altitude coordinates for each point in the world.
```

> To get actual gps readings we need to create a robot model with a GPS sensor. An updated turtle model is provided in the `tutorial repo <https://github.com/ros-planning/navigation2_tutorials/tree/master/nav2_gps_waypoint_follower_demo/models/turtlebot_waffle_gps>`_, which uses gazebo_ros's gps sensor plugin to output ``NavSatFix`` messages on the topic ``/gps/fix``.

Reviewer: Below this paragraph would be a great place to put a code block with the GPS code added to the URDF for reference

Reviewer: Imagine that the user is trying to set this up for their robot by reading this -- so we should provide the example changes where made inline so they can copy/paste for their particular situation

Build the nav2_gps_waypoint_follower_demo package, source your workspace and test your gazebo world is properly set up by launching: 

>   ``ros2 launch nav2_gps_waypoint_follower_demo gazebo_gps_world.launch.py``

Reviewer: Try to use actual code blocks, see the other tutorials for the rst formatting

Reviewer: ^ also across this file, there are alot of places this formatting appears

A turtlebot waffle should appear in the sonoma raceway world. You may also echo the topic ``/gps/fix`` to verify the robot is indeed producing gps measurements 

> .. image:: images/Gps_Navigation/gazebo_sonoma_raceway.png

Reviewer: What's up with this having no skin (e.g. just white)?

Author: I have looked into this and It looks like the model in nav2 has no skin either, at least it always shows white in my computer even in the turtlebot world simulation. I think this happens because the meshes in the [model](https://github.com/ros-planning/navigation2/blob/main/nav2_bringup/worlds/waffle.model) file don't have a `<material>` tag. Could you please check if the turtlebot3 model has a skin in your simulation in the turtlebot3 and let me know?

    :width: 700px
    :align: center
    :alt: Turtlebot in the sonoma raceway
 
1- Setup GPS Localization system
--------------------------------

> Once you have your simulation (or real robot) up and running, it's time to set up your localization system. Remember that nav2 needs a ``tf`` chain with the structure ``map`` -> ``odom`` -> ``base_link`` -> ``[sensor frames]``; global localization (``map`` -> ``base_link``) is usually provided by ``amcl``, while local odometry (``odom`` -> ``base_link``) is usually provided by the user's odometry system (wheel odometry, visual odometry, etc).

Reviewer: ```suggestion
Once you have your simulation (or real robot) up and running, it's time to set up your localization system. Remember that nav2 needs a ``tf`` chain with the structure ``map`` -> ``odom`` -> ``base_link`` -> ``[sensor frames]``; global localization (``map`` -> ``odom``) is usually provided by ``amcl``, while ``odom`` -> ``base_link`` is usually provided by the user's odometry system (wheel odometry, visual odometry, etc).
```

> In this tutorial, the GPS sensor on the robot will replace ``amcl`` in providing global localization. Though you may build a custom module that takes in the ``NavSatFix`` and ``Imu`` messages of your GPS and imu, and outputs a ``tf`` between your ``map`` and ``base_link`` frames using a local tangent plane, nav2's gps waypoint follower needs robot_localization to be used for this purpose. This package already has a node that performs the GPS -> local cartesian conversions called the `navsat_transform_node <http://docs.ros.org/en/jade/api/robot_localization/html/navsat_transform_node.html>`_, and features state estimation nodes that use Kalman Filters to fuse multiple sources of data.

Reviewer: map->odom. It might calculate map->base but we actually publish map->odom via TF

Reviewer: @pepisg missing

> In this tutorial, the GPS sensor on the robot will replace ``amcl`` in providing global localization. Though you may build a custom module that takes in the ``NavSatFix`` and ``Imu`` messages of your GPS and imu, and outputs a ``tf`` between your ``map`` and ``base_link`` frames using a local tangent plane, nav2's gps waypoint follower needs robot_localization to be used for this purpose. This package already has a node that performs the GPS -> local cartesian conversions called the `navsat_transform_node <http://docs.ros.org/en/jade/api/robot_localization/html/navsat_transform_node.html>`_, and features state estimation nodes that use Kalman Filters to fuse multiple sources of data.

Reviewer: Nav2 **uses** robot localization for this purpose, not "needs".

We will setup one extended kalman filter for local odometry, fusing wheel odometry and IMU data; and a second one for global localization, fusing the local cartesian coverted GPS coordinates, the wheel odometry and the IMU data. This is a common setup on robot_localization when using GPS data and more details around its configuration can be found in `RL's docs <http://docs.ros.org/en/jade/api/robot_localization/html/integrating_gps.html>`_. 

> A `configuration file <https://github.com/ros-planning/navigation2_tutorials/tree/master/nav2_gps_waypoint_follower_demo/config/dual_ekf_navsat_params.yaml>`_ and a `launch file <https://github.com/ros-planning/navigation2_tutorials/tree/master/nav2_gps_waypoint_follower_demo/launch/dual_ekf_navsat.launch.py>`_ are provided for this purpose. You may take a while before continuing to understand these two files and what they configure. Note that the EKFs are set to work in 2D mode. This is because nav2's costmap environment representation is 2-Dimensional, and several layers rely on the ``base_link`` frame being on the same plane as their global frame (``map`` or ``odom`` if global or local costmap) for the height related parameters to make sense.

Reviewer: I think it might be worth walking through some of this file, especially the NavSat Transform configuration. The Filters I wouldn't go crazy explaining but some introduction would be good

> The navsat transform node exposes the ``datum`` parameter to set the GPS coordinates of the origin of the local tangent plane, which given translates to the origin of the position measurements output through its odometry output (``odom1`` in RL's params). The node will set this automatically to the coordinates of the first valid `NavSatFix` message it gets, however you may specify them said parameter in the yaml file or calling the ``/datum`` service in runtime. In this tutorial we will go with the automatic initialization because there is no information about the environment stored in cartesian coordinates (a static map, semantic navigation waypoints, etc), however if that's the case in your application you may fix the ``datum`` so a given pair of coordinates produced by the GPS always correspond to the same cartesian coordinates in your reference system.

Reviewer: This first sentence is a bit run on and needs to be reworded. I'm having trouble understanding the second half of it

"however you may specify them said parameter in the yaml file" --> the?

"calling" -> "call"

"in runtime" to "at runtime"

> A `static transforms launch <https://github.com/ros-planning/navigation2_tutorials/tree/master/nav2_gps_waypoint_follower_demo/launch/static_transforms.launch.py>`_ is provided to set the transforms between ``base_link`` and all the robot sensors for nav2 and robot_localization.

Reviewer: This no longer exists, I don't think this paragraph is necessary anymore

As a sanity check that everything is working correctly, launch RL's launch file while gazebo is still running: 

  ``ros2 launch nav2_gps_waypoint_follower_demo dual_ekf_navsat.launch.py``

On a different terminal launch mapviz using the pre-built `config file <https://github.com/ros-planning/navigation2_tutorials/tree/master/nav2_gps_waypoint_follower_demo/config/gps_wpf_demo.mvc>`_ in the repo. `Get a bing maps API key <https://www.microsoft.com/en-us/maps/create-a-bing-maps-key>`_ and use it to display satellite pictures.

>   ``ros2 launch nav2_gps_waypoint_follower_demo mapviz.launch.py``

Reviewer: After this command, a picture of mapviz would be nice

And finally run the turtle teleop key node to teleoperate the simulated turtlebot: 

  ``ros2 run teleop_twist_keyboard teleop_twist_keyboard``

When you have everything up and running, start teleoperating the turtlebot and check that:

1. When the robot faces east (default initial heading) and you move it forward, the ``base_link`` frame (green arrow) moves east consistently with the raw GPS measurements (blue dot).

2. Movement is consistent overall not only when facing east, meaning that the GPS measurements are consistent with the robot heading and movement direction, and that they are consistent with the position of the robot in the world (for instance, when the robot moves towards the finish line, GPS measurements in mapviz do as well).

The gif below shows what you should see:

.. image:: images/Gps_Navigation/localization_check.gif
  :width: 600px
  :align: center

Sensors in a real robot may be less accurate than gazebo's, specially GPSs and absolute heading measurements from IMUs. To mitigate this you can leverage robot_localization's EKFs to complement sensor's capabilities:

> 1. If your IMU does not provide absolute heading measurements accurately, consider setting the ``differential`` parameter of its input to RL to ``true``. This way the filter will only fuse changes in the orientation and derive the absolute value from its motion model, internally differentiating changes in the absolute position to estimate where the robot was heading.

Reviewer: So how does this deal with the fact that the IMU never has the absolute orientation w.r.t. global coordinates and only on the initialized pose? Or does the EKF somehow above deal with that already?

2. If your GPS is noisy but you have a trustworthy wheel odometry source, consider tuning the sensors and process noise covariances to make the filter "trust" more or less one data source or its own internal state estimate. A properly tuned filter should be able to reject wrong GPS measurements to some degree.

2- Setup Navigation system
--------------------------

> Once you have your localization system up and running it's time to setup nav2. since RL is already providing the ``tf`` tree we don't need to launch ``amcl``, thus we can remove its parameters from the params file.

Reviewer: ```suggestion
Once you have your localization system up and running it's time to setup nav2. since RL is already providing the ``tf`` tree we don't need to launch ``amcl``, thus we can remove its parameters from the params file and node from Nav2 launch files.
```

> Outdoors environments can get quite big, to a degree that they may not me practically represented on a single costmap. For that reason in this tutorial we use a rolling global costmap that is big enough for fitting successive pairs of waypoints, however depending on your application you may still choose to use a fixed global costmap, just remember to make it fit all the potential locations the robot may visit.

Reviewer: may not "be practical"

> Outdoors environments can get quite big, to a degree that they may not me practically represented on a single costmap. For that reason in this tutorial we use a rolling global costmap that is big enough for fitting successive pairs of waypoints, however depending on your application you may still choose to use a fixed global costmap, just remember to make it fit all the potential locations the robot may visit.

Reviewer: Make a bulleted list of the options:
- Have a static map used to size the costmap for global planning
- Have a large but realistic environment and set the costmap size to the space's size 
- Have a very large environment so have a rolling costmap large enough to fit any consecutive tasks

Additionally we assume that there's no static map of the environment, however if you decide to use one make sure to maintain consistency with GPS coordinates by selecting a fixed datum.

We provide a `nav2 params file <https://github.com/ros-planning/navigation2_tutorials/tree/master/nav2_gps_waypoint_follower_demo/config/nav2_no_map_params.yaml>`_ and a `launch file <https://github.com/ros-planning/navigation2_tutorials/tree/master/nav2_gps_waypoint_follower_demo/gps_waypoint_follower.launch.py>`_ to put it all together. Remember that the GPS setup of robot_localization was just a mean for setting up the global localization system, however nav2 is still a cartesian navigation stack and you may still use all its cartesian tools. To confirm that everything is working, launch the provided file (this launches gazebo and RL as well so close them if you have them running from the previous steps) and use rviz to send a goal to the robot:

  ``ros2 launch nav2_gps_waypoint_follower_demo gps_waypoint_follower.launch.py``

> The gif below shows what you should see

Reviewer: ```suggestion
The gif below shows what you should see Nav2 navigating the robot autonomously!
```

.. image:: images/Gps_Navigation/navigation_check.gif
  :width: 600px
  :align: center

3-  Interactive GPS Waypoint Follower
-------------------------------------

> Now that we have performed our complete system setup, lets leverage nav2 GPS waypoint follower capabilities to navigate to goals that are expressed directly in GPS coordinates. For this demo we want to build an interactive interface similar to rviz's that allows us to click over a map to make the robot navigate to the clicked location. For that we will use mapviz's point click publisher on the ``wgs84`` reference frame, which will publish a ``PointStamped`` message with the GPS coordinates of the point clicked over the satellite image.

Reviewer: Mention that this is a useful demo to showcase how to get started

  ``ros2 run nav2_gps_waypoint_follower_demo interactive_waypoint_follower``

Then launch mapviz using the pre-built config:

  ``ros2 launch nav2_gps_waypoint_follower_demo mapviz.launch.py``

You can now click on the mapviz map the pose you want the robot to go. The gif below shows the robot navigating to the finish line going through some obstacles:

.. image:: images/Gps_Navigation/interactive_wpf.gif
  :width: 600px
  :align: center

4-  Logged GPS Waypoint Follower
--------------------------------

Finally let's make a robot go through a set of predefined GPS waypoints stored in a yaml file. For this purpose we provide the `logged_waypoint_follower <https://github.com/ros-planning/navigation2_tutorials/tree/master/nav2_gps_waypoint_follower_demo/nav2_gps_waypoint_follower_demo/logged_waypoint_follower.py>`_ node and a `waypoints <https://github.com/ros-planning/navigation2_tutorials/tree/master/nav2_gps_waypoint_follower_demo/config/demo_waypoints.yaml>`_ file. To test this node source your workspace and with the rest of the system running type:

>   ``ros2 run nav2_gps_waypoint_follower_demo logged_waypoint_follower``

Reviewer: Needs gif / graphic 

Also after this section have a short conclusion section to round it off

.. image:: images/Gps_Navigation/South-America-UTM-zones.png
    :width: 520px
    :align: center
    :alt: UTM grid zones in South America

`robot_localization <http://docs.ros.org/en/noetic/api/robot_localization/html/index.html>`_ uses this projection system to transform GPS measurements in the WGS84 reference system to a cartesian system, which centered on the origin of the grid zone where the GPS is at. This is achieved through the `navsat_transform node <http://docs.ros.org/en/jade/api/robot_localization/html/navsat_transform_node.html>`_. This node complies with the ENU convention in `REP 103 <https://www.ros.org/reps/rep-0103.html>`_, meaning that the ``+x`` axis of the ``utm`` coordinate system faces east, the ``+y`` faces north and the ``+z`` axis points up.

> In the real world GPS sensors can be noisy: With standalone GPSs you should expect accuracies of 1-2 meters under excellent conditions and up to 10 meters, and frequent jumps in the position as the GPS sensor picks up less or more satellites, which can degrade the quality of navigation significantly. Several positioning augmentation technologies exists to reduce the error of GPS measurements, one of the most common ones is called `RTK <https://en.wikipedia.org/wiki/Real-time_kinematic_positioning>`_ (Real Time Kinematic Positioning), which can bring the accuracy of receivers down to 1cm. If accuracy matters in your application this technology is highly recommended; though this requires the deployment of a second fixed GPS called base, most of the US and Europe are already covered with public free to use bases that you can connect to. You can read more about RTK and how to get started `here <https://learn.sparkfun.com/tutorials/setting-up-a-rover-base-rtk-system>`_. In this tutorial we assume the robot's GPS produces a really accurate and smooth estimation of the robot's position.

Reviewer: ```suggestion
In the real world GPS sensors can be noisy: With standalone GPSs you should expect accuracies of 1-2 meters under excellent conditions and up to 10 meters, and frequent jumps in the position as the GPS sensor picks up less or more satellites, which can degrade the quality of navigation significantly. Several positioning augmentation technologies exists to reduce the error of GPS measurements, one of the most common ones is called `RTK <https://en.wikipedia.org/wiki/Real-time_kinematic_positioning>`_ (Real Time Kinematic Positioning), which can bring the accuracy of receivers down to 1cm. If accuracy matters in your application this technology is highly recommended; though this requires the deployment of a second fixed GPS called base, most of the US and Europe are already covered with public free to use bases that you can connect to. You can read more about RTK and how to get started `here <https://learn.sparkfun.com/tutorials/setting-up-a-rover-base-rtk-system>`_. In this tutorial we assume the robot's GPS produces an accurate and smooth estimation of the robot's position.
```

  <joint name="base_joint" type="fixed">
    <parent link="base_link"/>
    <child link="base_footprint" />
    <origin xyz="0 0 -0.010" rpy="0 0 0"/>
  </joint>

Build the ``nav2_gps_waypoint_follower_demo`` package, source your workspace and test your gazebo world is properly set up by launching: 

.. code-block:: bash

  ros2 launch nav2_gps_waypoint_follower_demo gazebo_gps_world.launch.py

A Turtlebot waffle should appear in the Sonoma Raceway world. You may also echo the topic ``/gps/fix`` to verify the robot is indeed producing GPS measurements 

.. image:: images/Gps_Navigation/gazebo_sonoma_raceway.png
    :width: 700px
    :align: center
    :alt: Turtlebot in the sonoma raceway
 
1- Setup GPS Localization system
--------------------------------

Once you have your simulation (or real robot) up and running, it's time to set up your localization system. Remember that Nav2 uses a ``tf`` chain with the structure ``map`` -> ``odom`` -> ``base_link`` -> ``[sensor frames]``; global localization (``map`` -> ``odom``) is usually provided by ``amcl``, while ``odom`` -> ``base_link`` is usually provided by the user's odometry system (wheel odometry, visual odometry, etc).

> In this tutorial, the GPS sensor on the robot will replace ``amcl`` in providing global localization. Though you may build a custom module that takes in the ``NavSatFix`` and ``Imu`` messages of your GPS and imu, and outputs a ``tf`` between your ``map`` and ``odom`` frames using a planar projection, Nav2's GPS waypoint follower currently uses robot_localization for converting GPS goals to cartesian goals, and thus at a `navsat_transform_node <http://docs.ros.org/en/jade/api/robot_localization/html/navsat_transform_node.html>`_ shoud be active. Additionally, ``robot_localization`` features reconfigurable state estimation nodes that use Kalman Filters to fuse multiple sources of data, which is yet another reason to use it.

Reviewer: ```suggestion
In this tutorial, the GPS sensor on the robot will replace ``amcl`` in providing global localization. Though you may build a custom module that takes in the ``NavSatFix`` and ``Imu`` messages of your GPS and imu, and outputs a ``tf`` between your ``map`` and ``odom`` frames using a planar projection, Nav2's GPS waypoint follower currently uses robot_localization for converting GPS goals to cartesian goals, and thus at a `navsat_transform_node <http://docs.ros.org/en/jade/api/robot_localization/html/navsat_transform_node.html>`_ should be active. Additionally, ``robot_localization`` features reconfigurable state estimation nodes that use Kalman Filters to fuse multiple sources of data, which is yet another reason to use it.
```

.. _navigation2-with-gps:

Navigating Using GPS Localization
*********************************

- `Overview`_
- `Requirements`_
- `GPS Localization Overview`_
- `Tutorial Steps`_

.. image:: images/Gps_Navigation/interactive_wpf.gif
  :width: 600px
  :align: center

Overview
========

> This tutorial shows how to set up a localization system using a GPS sensor as source of global positioning and robot_localization (RL) for sensor fusion, and how to use Nav2 to follow GPS waypoints. It was written by Pedro Gonzalez at `Kiwibot <https://www.kiwibot.com/>`_

Reviewer: ```suggestion
This tutorial shows how to set up a localization system using a GPS sensor(s) as the source of global positioning, robot_localization (RL) for sensor fusion, and how to use Nav2 to follow GPS waypoints. It was written by Pedro Gonzalez at `Kiwibot <https://www.kiwibot.com/>`_.
```

.. image:: images/Gps_Navigation/South-America-UTM-zones.png
    :width: 520px
    :align: center
    :alt: UTM grid zones in South America

`robot_localization <http://docs.ros.org/en/noetic/api/robot_localization/html/index.html>`_ uses this projection system to transform GPS measurements in the WGS84 reference system to a cartesian system, which centered on the origin of the grid zone where the GPS is at. This is achieved through the `navsat_transform node <http://docs.ros.org/en/jade/api/robot_localization/html/navsat_transform_node.html>`_. This node complies with the ENU convention in `REP 103 <https://www.ros.org/reps/rep-0103.html>`_, meaning that the ``+x`` axis of the ``utm`` coordinate system faces east, the ``+y`` faces north and the ``+z`` axis points up.

> In this tutorial we assume the robot's GPS produces a really accurate and smooth estimation of the robot's position, however in the real world for standalone GPSs that's often not the case: you should expect accuracies of 1-2 meters under excellent conditions and up to 10 meters, and frequent jumps in the position as the GPS sensor picks up less or more satellites.

Reviewer: This paragraph has good context but its a bit misleading up the front. The first sentence makes it sound like there's some critical problem that this tutorial is just ignoring. But then its clear later on in the paragraph there are common solutions. 

I think the first sentence should be re-written. Maybe ignore the "In this tutorial we assume..." bit and just explain the GPS jumps and here are some solutions more abstractly. Maybe at the end of this paragraph we just say "In this tutorial, we assume we have adequate localization via RTK or similar technologies or good conditions".

Several positioning augmentation technologies exists to reduce the error of GPS measurements, one of the most common ones is called `RTK <https://en.wikipedia.org/wiki/Real-time_kinematic_positioning>`_ (Real Time Kinematic Positioning), which can bring the accuracy of receivers down to 1cm. If accuracy matters in your application this technology is highly recommended; though this requires the deployment of a second fixed GPS called base, most of the US and Europe are already covered with public free to use bases that you can connect to. You can read more about RTK and how to get started on `this tutorial <https://learn.sparkfun.com/tutorials/setting-up-a-rover-base-rtk-system>`_.

> Additionally, to fully describe a robot's localization we need to know its heading as well, however standalone GPS sensors do not provide orientation measurements, only position measurements. In this tutorial we will refer as absolute heading to a yaw measurement which is given w.r.t. a cardinal direction (e.g, the east), in contrast to relative heading, which is given w.r.t. the angle the robot is turned on or any other reference that cannot be directly mapped  to a cardinal direction.

Reviewer: ```suggestion
Additionally, to fully describe a robot's localization we need to know its heading as well, however standalone GPS sensors do not provide orientation measurements, only position measurements. In this tutorial we will refer as 'absolute heading' to a yaw measurement which is given w.r.t. a cardinal direction (e.g, the east), in contrast to relative heading, which is given w.r.t. the angle the robot is turned on or any other reference that cannot be directly mapped  to a cardinal direction.
```

> When using robot_localization with GPS, measuring absolute orientation is mandatory. There are several alternatives for getting absolute orientation data, like IMUs with magnetometers, dual GPS systems or matching techniques over a known map; in this tutorial we assume the robot is equipped with an IMU that can accurately measure absolute orientation following the ENU convention, meaning it will output zero yaw when facing east and +90 degrees when facing north. 

Reviewer: ```suggestion
When using robot_localization with GPS, measuring absolute orientation is mandatory. There are several strategies for getting absolute orientation data, like IMUs with magnetometers, dual GPS systems or matching techniques over a known map; in this tutorial we assume the robot is equipped with an IMU that can accurately measure absolute orientation following the ENU convention, meaning it will output zero yaw when facing east and +90 degrees when facing north. 
```

Despite the above assumption, in the real world commercial grade IMU's mounted in actual robots will often not produce accurate absolute heading measurements because: 

1. They may not have a magnetometer.

2. They are hard to calibrate: outdoors robots are often big and heavy: imagine doing an eight figure in the air with an autonomous tractor.

3. Robots can be a huge source of electromagnetic noise for magnetometers: Electric motors are full of permanent magnets and can draw several amps, producing significant disturbances to the sensor.

> Through the development of the tutorial we will see how to leverage robot_localization's Kalman Filters to mitigate this problem.

Reviewer: ```suggestion
Thus, for a particular application you should consider the behavior and localization quality you require when making decisions about how to estimate your absolute heading. When using IMU's without relative headings to a cardinal direction, the robot may need to move around for a bit in an 'initialization dance' to converge to the right heading using the filter. Using dual-GPS or 3D mapping system overlay, the initial heading is quite good. 

For the purposes of this tutorial, we model a well-built system using an IMU that has absolute orientation already, but that may be augmented or replaced on a practical system using one of the techniques above (or others).
```

  <joint name="base_joint" type="fixed">
    <parent link="base_link"/>
    <child link="base_footprint" />
    <origin xyz="0 0 -0.010" rpy="0 0 0"/>
  </joint>

Build the nav2_gps_waypoint_follower_demo package, source your workspace and test your gazebo world is properly set up by launching: 

.. code-block:: bash

  ros2 launch nav2_gps_waypoint_follower_demo gazebo_gps_world.launch.py

A Turtlebot waffle should appear in the Sonoma Raceway world. You may also echo the topic ``/gps/fix`` to verify the robot is indeed producing GPS measurements 

.. image:: images/Gps_Navigation/gazebo_sonoma_raceway.png
    :width: 700px
    :align: center
    :alt: Turtlebot in the sonoma raceway
 
1- Setup GPS Localization system
--------------------------------

Once you have your simulation (or real robot) up and running, it's time to set up your localization system. Remember that Nav2 uses a ``tf`` chain with the structure ``map`` -> ``odom`` -> ``base_link`` -> ``[sensor frames]``; global localization (``map`` -> ``odom``) is usually provided by ``amcl``, while ``odom`` -> ``base_link`` is usually provided by the user's odometry system (wheel odometry, visual odometry, etc).

> In this tutorial, the GPS sensor on the robot will replace ``amcl`` in providing global localization. Though you may build a custom module that takes in the ``NavSatFix`` and ``Imu`` messages of your GPS and imu, and outputs a ``tf`` between your ``map`` and ``base_link`` frames using a planar projection, Nav2's GPS waypoint follower needs robot_localization to be used. As mentioned above, This package already has the `navsat_transform_node <http://docs.ros.org/en/jade/api/robot_localization/html/navsat_transform_node.html>`_ for this purpose, and features state estimation nodes that use Kalman Filters to fuse multiple sources of data.

Reviewer: Remove any reference to Nav2 "needs" R_L, we don't "need" it, we're currently using it. There are other solutions as well, including the ones we're working on recently with Fuse

> In this tutorial, the GPS sensor on the robot will replace ``amcl`` in providing global localization. Though you may build a custom module that takes in the ``NavSatFix`` and ``Imu`` messages of your GPS and imu, and outputs a ``tf`` between your ``map`` and ``base_link`` frames using a planar projection, Nav2's GPS waypoint follower needs robot_localization to be used. As mentioned above, This package already has the `navsat_transform_node <http://docs.ros.org/en/jade/api/robot_localization/html/navsat_transform_node.html>`_ for this purpose, and features state estimation nodes that use Kalman Filters to fuse multiple sources of data.

Author: But the GPS waypoint follower does "need" robot localization to be used (or at least the navsat_transform node) because it relies on the `/fromLL` service to convert GPS to cartesian goals. Maybe I could add the reason why its needed so its clearer

> In this tutorial, the GPS sensor on the robot will replace ``amcl`` in providing global localization. Though you may build a custom module that takes in the ``NavSatFix`` and ``Imu`` messages of your GPS and imu, and outputs a ``tf`` between your ``map`` and ``base_link`` frames using a planar projection, Nav2's GPS waypoint follower needs robot_localization to be used. As mentioned above, This package already has the `navsat_transform_node <http://docs.ros.org/en/jade/api/robot_localization/html/navsat_transform_node.html>`_ for this purpose, and features state estimation nodes that use Kalman Filters to fuse multiple sources of data.

Reviewer: That another option could provide that as well. As you know, we're moving GPS / state estimation to Fuse right now, so clearly R_L is not going to be the only game in town as a requirement

.. code-block:: yaml

  ekf_filter_node_odom:
    ros__parameters:
      two_d_mode: true
      publish_tf: true

      base_link_frame: base_link
      world_frame: odom

Since per `REP 105 <https://www.ros.org/reps/rep-0105.html>`_ the position of the robot in the ``odom`` frame has to be continuous over time, in this filter we just want to fuse the robot's speed measured by its wheels published ``/odom``, and the imu heading published on ``/imu``:

.. code-block:: yaml

  odom0: odom
  odom0_config: [false, false, false,
                false, false, false,
                true,  true,  true,
                false, false, true,
                false, false, false]

  imu0: imu
  imu0_config: [false, false, false,
                false,  false,  true,
                false, false, false,
                false,  false,  false,
                false,  false,  false]

Global Odometry
^^^^^^^^^^^^^^^

> The global odometry is provided by the ``ekf_filter_node_map``, which publishes the transform between ``map`` and ``base_link``. This EKF is set to work in 2D mode as well. In addition to the IMU and wheel odometry data, this filter takes in the odometry output of the gps, published by the ``navsat_transform`` node on ``/odometry/gps``:

Reviewer: ```suggestion
The global odometry is provided by the ``ekf_filter_node_map``, which publishes the transform between ``map`` and ``base_link``. This EKF is set to work in 2D mode as well. In addition to the IMU and wheel odometry data, this filter takes in the odometry output of the gps, published by the ``navsat_transform`` node on ``/odometry/gps``, as an odometry message:
```

1. **Rolling** (Used in the tutorial): Outdoors environments can get quite big, to a degree that it may not be practical to represent them on a single costmap. For that reason in this tutorial we use a rolling global costmap that is big enough for fitting successive pairs of waypoints. In this case you may or may not choose to use a static layer, however if you do make sure to fix the ``datum`` of the navsat_transform so GPS coordinates always have the same cartesian representation on your map.

.. code-block:: yaml

  global_costmap:
    global_costmap:
      ros__parameters:
        ...
        rolling_window: True
        width: 50
        height: 50

2. **Size and position from static map**: You may also choose to keep Nav2 default setup and have the global costmap be sized and positioned according to a pre-built map by adding a static layer and using ``map_server``. In this case you also need to make sure there's consistency in your ``datum`` and the origin of the map.

.. code-block:: yaml

  global_costmap:
    global_costmap:
      ros__parameters:
        ...
        plugins: ["static_layer", "obstacle_layer", "inflation_layer"]

> 3. **Static position and size**: Finally, depending on your application you may still choose to use a fixed global costmap, just remember to make it fit all the potential locations the robot may visit. In this case you need to set the size and origin position in the parameters:

Reviewer: ```suggestion
3. **Static position and size**: Finally, depending on your application you may still choose to use a fixed global costmap if you have a restricted operating environment you know beforehand, just remember to make it fit all the potential locations the robot may visit. In this case you need to set the size and origin position in the parameters:
```

After that you should get a ``yaml`` file in the location you specified with the format shown above; let's now make the robot follow the logged waypoints. For this purpose we provide the `logged_waypoint_follower <https://github.com/ros-planning/navigation2_tutorials/tree/master/nav2_gps_waypoint_follower_demo/nav2_gps_waypoint_follower_demo/logged_waypoint_follower.py>`_ node, which takes in the path to the waypoints file as an argument and uses the ``BasicNavigator`` in ``nav2_simple_commander`` to send the logged goals to the ``/follow_gps_waypoints`` action server. If not provided, the node uses the `default waypoints <https://github.com/ros-planning/navigation2_tutorials/tree/master/nav2_gps_waypoint_follower_demo/config/demo_waypoints.yaml>`_ in the ``nav2_gps_waypoint_follower_demo`` package.

To run this node source your workspace and with the rest of the system running type:

.. code-block:: bash

  ros2 run nav2_gps_waypoint_follower_demo logged_waypoint_follower </path/to/yaml/file.yaml>

You should now see the robot following the waypoints you previously logged: 

.. image:: images/Gps_Navigation/logged_waypoint_follower.gif
  :width: 800px
  :align: center

> 

Reviewer: Create a short conclusion section just to go over what we setup and conclude the tutorial
